### PR TITLE
api: support get releases endpoint

### LINKED ISSUE #6012: Support API: `GET /repos/:user/:repo/releases`
The API is currently missing the releases endpoint at `/repos/:user/:repo/releases`, at least in its basic form exposing the list of releases.

See https://developer.github.com/v3/repos/releases/#list-releases-for-a-repository

**Additional context**

This prevents use of Goreleaser with Gogs, although it is available in Gitea, causing Gitea to be supported there but not Gogs.

### PR DESCRIPTION
**Pull request by fishbullet**
Fixes #6012

Added the releases endpoint at `/repos/:user/:repo/releases`

https://developer.github.com/v3/repos/releases/#list-releases-for-a-repository

### CODE REVIEW

File: internal/route/api/v1/repo/repo.go [hunk 1]
```diff
@@ -403,3 +403,19 @@ func MirrorSync(c *context.APIContext) {
 	go db.MirrorQueue.Add(repo.ID)
 	c.Status(http.StatusAccepted)
 }
+
+func Releases(c *context.APIContext) {
+	_, repo := parseOwnerAndRepo(c)
+	releases, err := db.GetReleasesByRepoID(repo.ID)
+	if err != nil {
+		c.Error(err, "get releases")
```
Comment (unknwon) [inline internal/route/api/v1/repo/repo.go]:
```suggestion
		c.Error(err, "get releases by repository ID")
```

**Comment by fishbullet** [comment]
> Thanks for the PR!
> 
> Can you post example request input and output for this new endpoint? I think it might panic.

You're right, it's panic on the `Publisher` attributes.

I've added a publisher fetching like below:

```golang
func Releases(c *context.APIContext) {
	_, repo := parseOwnerAndRepo(c)
	releases, err := db.GetReleasesByRepoID(repo.ID)
	if err != nil {
		c.Error(err, "get releases by repository ID")
		return
	}
	apiReleases := make([]*api.Release, len(releases))
	for _, r := range releases {
		publisher, err := db.GetUserByID(r.PublisherID)
		if err != nil {
			c.Error(err, "get releases by repository ID cant find publisher")
			return
		}
		r.Publisher = publisher
	}
	for _, r := range releases {
		apiReleases = append(apiReleases, r.APIFormat())
	}

	c.JSONSuccess(&apiReleases)
}
```

And now there is additional null entries in the response:

```json
~ > curl -H "Authorization: token b75b4d4e5138f71964764c079da7f58e04ea2fa3" 'http://localhost:3000/api/v1/repos/foo/ddd/releases' | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   339  100   339    0     0   8071      0 --:--:-- --:--:-- --:--:--  8071
[
  null,
  {
    "id": 2,
    "tag_name": "fff",
    "target_commitish": "master",
    "name": "dwad",
    "body": "",
    "draft": false,
    "prerelease": false,
    "author": {
      "id": 1,
      "username": "foo",
      "login": "foo",
      "full_name": "",
      "email": "foo@bar.com",
      "avatar_url": "https://secure.gravatar.com/avatar/f3ada405ce890b6f8204094deb12d8a8?d=identicon"
    },
    "created_at": "2020-03-28T13:38:00+03:00"
  }
]
~ 
```

Can't get where it is come from.

**Comment by unknwon** [comment]
`apiReleases := make([]*api.Release, 0, len(releases))`

**Comment by fishbullet** [comment]
> `apiReleases := make([]*api.Release, 0, len(releases))`

My bad :confused:

**Comment by fishbullet** [comment]
@unknwon Do I need to make a changes in [documentation](https://github.com/gogs/docs-api/tree/master/Repositories) for this API ?

**Comment by unknwon** [comment]
> @unknwon Do I need to make a changes in [documentation](https://github.com/gogs/docs-api/tree/master/Repositories) [](/gogs/docs-api/tree/HEAD@{2020-03-28T13:29:01Z}/Repositories) for this API ?

Yes, but not until this PR is merged.

File: internal/db/release.go [hunk 2]
```diff
@@ -267,6 +267,12 @@ func GetPublishedReleasesByRepoID(repoID int64, matches ...string) ([]*Release,
 	return releases, sess.Find(&releases, new(Release))
 }
 
+// GetReleasesByRepoID returns a list of all releases of repository
```
Comment (unknwon) [inline internal/db/release.go]:
```suggestion
// GetReleasesByRepoID returns a list of all releases (including drafts) of given repository.
```

File: internal/route/api/v1/repo/repo.go [hunk 3]
```diff
@@ -403,3 +403,26 @@ func MirrorSync(c *context.APIContext) {
...
+	releases, err := db.GetReleasesByRepoID(repo.ID)
+	if err != nil {
+		c.Error(err, "get releases by repository ID")
+		return
+	}
+	apiReleases := make([]*api.Release, 0, len(releases))
+	for _, r := range releases {
+		publisher, err := db.GetUserByID(r.PublisherID)
+		if err != nil {
+			c.Error(err, "get releases by repository ID cant find publisher")
```
Comment (unknwon) [inline internal/route/api/v1/repo/repo.go]:
```suggestion
			c.Error(err, "get release publisher")
```

**Comment by unknwon** [comment]
FYI, you can commit suggestions directly in GitHub UI.

**Comment by fishbullet** [comment]
> FYI, you can commit suggestions directly in GitHub UI.

Yeah, I know. 
I like clear pull commits message, so I prefer do a squash.

**Comment by unknwon** [comment]
> > FYI, you can commit suggestions directly in GitHub UI.
> 
> Yeah, I know.
> I like clear pull commits message, so I prefer do a squash.

We always do squash and merge, so no need to do that manually anymore :)

**Comment by unknwon** [comment]
Merging, congrats!

**Pull request merged**